src/github.com/matrix-org/dendrite/clientapi/writers/register.go

```diff
@@ -0,0 +1,153 @@
+	// https://github.com/matrix-org/synapse/blob/v0.20.0/synapse/rest/client/v2_alpha/register.py#L161
+	if len(r.Password) > 512 {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.BadJSON("'password' >512 characters"),
+		}
+	} else if len(r.Username) > 512 {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.BadJSON("'username' >512 characters"),
+		}
+	} else if len(r.Password) > 0 && len(r.Password) < 8 {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.WeakPassword("password too weak: min 8 chars"),
+		}
+	}
+	return nil
+}
+
+// Register processes a /register request. http://matrix.org/speculator/spec/HEAD/client_server/unstable.html#post-matrix-client-unstable-register
+func Register(req *http.Request, accountDB *storage.AccountDatabase) util.JSONResponse {
+	var r registerRequest
+	resErr := httputil.UnmarshalJSONRequest(req, &r)
+	if resErr != nil {
+		return *resErr
+	}
+	if resErr = r.Validate(); resErr != nil {
+		return *resErr
+	}
+
+	logger := util.GetLogger(req.Context())
+	logger.WithFields(log.Fields{
+		"username":   r.Username,
+		"auth.type":  r.Auth.Type,
+		"session_id": r.Auth.Session,
+	}).Info("Processing registration request")
+
+	// TODO: Shared secret registration (create new user scripts)
+	// TODO: AS API registration
+	// TODO: Enable registration config flag
+	// TODO: Guest account upgrading
+
+	// All registration requests must specify what auth they are using to perform this request
+	if r.Auth.Type == "" {
+		return util.JSONResponse{
+			Code: 401,
+			// TODO: Hard-coded 'dummy' auth for now with a bogus session ID.
+			//       Server admins should be able to change things around (eg enable captcha)
+			JSON: newUserInteractiveResponse("totallyuniquesessionid", []flow{
+				flow{[]types.LoginType{types.LoginTypeDummy}},
+			}),
+		}
+	}
+
+	// TODO: Handle loading of previous session parameters from database.
+	// TODO: Handle mapping registrationRequest parameters into session parameters
+
+	// TODO: email / msisdn / recaptcha auth types.
+	switch r.Auth.Type {
+	case types.LoginTypeDummy:
+		// there is nothing to do
+		return completeRegistration(accountDB, r.Username, r.Password)
+	default:
+		return util.JSONResponse{
+			Code: 501,
+			JSON: jsonerror.Unknown("unknown/unimplemented auth type"),
+		}
+	}
+}
+
+func completeRegistration(accountDB *storage.AccountDatabase, username, password string) util.JSONResponse {
+	acc, err := accountDB.CreateAccount(username, password)
+	if err != nil {
+		return util.JSONResponse{
+			Code: 500,
+			JSON: jsonerror.Unknown("failed to create account: " + err.Error()),
+		}
+	}
+	// TODO: Make and store a proper access_token
+	// TODO: Store the client's device information?
+	return util.JSONResponse{
+		Code: 200,
+		JSON: registerResponse{
+			UserID:      acc.UserID,
+			AccessToken: acc.UserID, // FIXME
+			HomeServer:  acc.ServerName,
+			DeviceID:    "dendrite",
```

Comment:
It could be clearer that this is a dummy place-holder value, maybe "kindauniquedeviceid" or something?